model/vm_host.rb

```diff
@@ -243,16 +244,70 @@ def reset
     Hosting::Apis.reset_server(self)
   end
 
+  def check_storage_smartctl(ssh_session, devices)
+    devices.map { |device_name| ssh_session.exec!("sudo smartctl -j -H /dev/#{device_name} | jq .smart_status.passed").strip == "true" }.all?(true)
+  end
+
+  def check_storage_nvme(ssh_session, devices)
+    devices.map { |device_name| ssh_session.exec!("sudo nvme smart-log /dev/#{device_name} | grep \"critical_warning\" | awk '{print $3}'").strip == "0" }.all?(true)
+  end
+
+  def check_storage_read_write(ssh_session, devices)
+    lsblk_json_info = ssh_session.exec!("lsblk --json")
+    device_mount_points = devices.map { |device| SystemParser.get_device_mount_points_from_lsblk(lsblk_json_info, device) }
+
+    all_mount_points = []
+    device_mount_points.each do |device_mount_point|
+      device_mount_point.each do |device_name, mount_points|
+        mount_points.each do |mount_point|
+          all_mount_points.append(mount_point)
+        end
+      end
+    end
+
+    success = true
+    all_mount_points.uniq.each do |mount_point|
+      file_name = (mount_point == "/") ? "/test-file" : "#{mount_point}/test-file"
+      ssh_session.exec!("sudo bash -c \"head -c 1M </dev/zero > #{file_name}\"")
+      success = false unless ssh_session.exec!("sha256sum #{file_name}").strip == "30e14955ebf1352266dc2ff8067e68104607e750abb9d3b36582b8af909fcb58  #{file_name}"
+      ssh_session.exec!("sudo rm #{file_name}")
+    end
+    success
+  end
+
+  def check_storage_kernel_logs(ssh_session, devices)
+    kernel_logs = ssh_session.exec!("journalctl -kS -1min --no-pager").split("\n")
+    error_count = Hash.new(0)
+    kernel_logs.reduce(error_count) do |error_count, line|
```

Comment:
Can we simplify this by just doing a regex search without reduce?

Reply:
done